### PR TITLE
Output last 20 lines of docker logs to stdout on failure

### PR DESCRIPTION
**Pull request by b-deam**
With this commit we print the last 20 lines when a container created by
`DockerBuilder` fails.

This brings the behaviour in line with the existing native `Builder`.

<details>

<summary> Example </summary>

Before:
```shell
$ esrally build --revision "@2022-02-02" --elasticsearch-plugins repository-s3,repository-gcs,repository-azure --source-build-method docker
    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Using Docker to create installable binary from source files
[ERROR] Cannot build. Docker container [esrally-source-builder] failed with status code [143]: Error [None]

Getting further help:
*********************
* Check the log files in /Users/bradleydeam/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/bradleydeam/.rally/logs.

--------------------------------
[INFO] FAILURE (took 28 seconds)
--------------------------------
```

After:
```shell
$ esrally build --revision "@2022-02-02" --elasticsearch-plugins repository-s3,repository-gcs,repository-azure --source-build-method docker

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Using Docker to create installable binary from source files
[ERROR] Cannot build. Docker container [esrally-source-builder] failed with status code [143]: Error [None]: Build log output [Executing 'esrally-source-builder' failed. The last 20 lines in the build log file are:
=========================================================================================================
        Get:7 http://ports.ubuntu.com/ubuntu-ports jammy/multiverse arm64 Packages [224 kB]
        Get:8 http://ports.ubuntu.com/ubuntu-ports jammy/universe arm64 Packages [17.2 MB]
        Get:9 http://ports.ubuntu.com/ubuntu-ports jammy-updates/universe arm64 Packages [817 kB]
        Get:10 http://ports.ubuntu.com/ubuntu-ports jammy-updates/multiverse arm64 Packages [2,308 B]
        Get:11 http://ports.ubuntu.com/ubuntu-ports jammy-updates/main arm64 Packages [848 kB]
        Get:12 http://ports.ubuntu.com/ubuntu-ports jammy-updates/restricted arm64 Packages [222 kB]
        Get:13 http://ports.ubuntu.com/ubuntu-ports jammy-backports/main arm64 Packages [3,184 B]
        Get:14 http://ports.ubuntu.com/ubuntu-ports jammy-backports/universe arm64 Packages [7,266 B]
        Get:15 http://ports.ubuntu.com/ubuntu-ports jammy-security/universe arm64 Packages [632 kB]
        Get:16 http://ports.ubuntu.com/ubuntu-ports jammy-security/main arm64 Packages [553 kB]
        Get:17 http://ports.ubuntu.com/ubuntu-ports jammy-security/restricted arm64 Packages [209 kB]
        Fetched 23.1 MB in 7s (3,154 kB/s)
        Reading package lists...
        Building dependency tree...
        Reading state information...
        23 packages can be upgraded. Run 'apt list --upgradable' to see them.
        groupadd: group 'staff' already exists
        /bin/bash: line 1: git: command not found
        Downloading https://services.gradle.org/distributions/gradle-7.3.3-all.zip
        ...............10%...............20%...............30%...............40%...............50%...............60%................70%...............80%....Downloading https://services.gradle.org/distributions/gradle-7.3.3-all.zip
=========================================================================================================
The full build log is available at [/Users/bradleydeam/.rally/logs/build.log].]

Getting further help:
*********************
* Check the log files in /Users/bradleydeam/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/bradleydeam/.rally/logs.

--------------------------------
[INFO] FAILURE (took 25 seconds)
--------------------------------
```

</details>

### CODE REVIEW

File: esrally/mechanic/supplier.py [hunk 1]
```diff
@@ -865,8 +865,16 @@ def tail_container_logs(self, output, container_name):
 
     def check_container_return_code(self, completion, container_name):
         if completion["StatusCode"] != 0:
-            raise SystemSetupError(
-                f"Docker container [{container_name}] failed with status code [{completion['StatusCode']}]: Error [{completion['Error']}]"
+            msg = "Executing '{}' failed. The last 20 lines in the build log file are:\n".format(container_name)
```
Comment (dliappis) [inline esrally/mechanic/supplier.py:868]: Use an f-string?

File: esrally/mechanic/supplier.py [hunk 2]
```diff
@@ -865,8 +865,16 @@ def tail_container_logs(self, output, container_name):
...
+            msg += "=========================================================================================================\n"
+            with open(self.log_file, "r", encoding="utf-8") as f:
+                msg += "\t"
+                msg += "\t".join(f.readlines()[-20:])
+            msg += "=========================================================================================================\n"
+            msg += "The full build log is available at [{}]".format(self.log_file)
```
Comment (dliappis) [inline esrally/mechanic/supplier.py:874]: same comment, let's stick to f-strings

File: tests/mechanic/supplier_test.py [hunk 3]
```diff
@@ -353,15 +353,19 @@ def docker_log_output(text):
...
         completion = {"StatusCode": 0, "Error": None}
         builder.check_container_return_code(completion, "test-container-name")
         assert "Container [test-container-name] completed successfully." in caplog.text
 
         completion = {"StatusCode": 1, "Error": "Vague error message"}
-        with pytest.raises(exceptions.SystemSetupError) as e:
-            builder.check_container_return_code(completion, "test-container-name")
-        assert "Docker container [test-container-name] failed with status code [1]: Error [Vague error message]" in str(e.value)
+        with mock.patch("builtins.open", mock.mock_open()) as file:
+            file.return_value.readlines.return_value = "my error"
```
Comment (dliappis) [inline tests/mechanic/supplier_test.py:362]: Which is this needed, give that the [check](https://github.com/elastic/rally/pull/1629/files#diff-bcbcac752601b8c90ba4cc053a13169690cea0da315ea09ced558fde60727adcR867) is `if completion["StatusCode"] != 0`?

**Review by dliappis — COMMENTED** [review]
Very useful! Left a few minor questions.

File: tests/mechanic/supplier_test.py (hunk 3 above)
Comment (b-deam) [reply tests/mechanic/supplier_test.py]: Leftover from my initial approach to regex for the full error message (but ran into complications with escaping characters). Removed in https://github.com/elastic/rally/pull/1629/commits/754598d65fd1be731787c58467d8f8234a7fba84

File: esrally/mechanic/supplier.py (hunk 1 above)
Comment (b-deam) [reply esrally/mechanic/supplier.py]: Addressed in https://github.com/elastic/rally/pull/1629/commits/efd06ea3e9ba685e87015815d7e8827176bb0b28

File: esrally/mechanic/supplier.py (hunk 2 above)
Comment (b-deam) [reply esrally/mechanic/supplier.py]: Addressed in https://github.com/elastic/rally/pull/1629/commits/efd06ea3e9ba685e87015815d7e8827176bb0b28

File: esrally/mechanic/supplier.py [hunk 4]
```diff
@@ -865,8 +865,16 @@ def tail_container_logs(self, output, container_name):
 
     def check_container_return_code(self, completion, container_name):
         if completion["StatusCode"] != 0:
-            raise SystemSetupError(
-                f"Docker container [{container_name}] failed with status code [{completion['StatusCode']}]: Error [{completion['Error']}]"
+            msg = f"Executing '{container_name}' failed. The last 20 lines in the build log file are:\n"
```
Comment (pquentin) [inline esrally/mechanic/supplier.py:868]:
```suggestion
            msg = f"Executing '{container_name}' failed. The last 20 lines in the build.log file are:\n"
```

**Review by pquentin — APPROVED** [review]
Thanks! LGTM. Left a nit.

**Review by dliappis — APPROVED** [review]
LGTM

**Comment by b-deam** [comment]
@elasticmachine run rally/it-python310

**Pull request merged**